rpm/0003-Use-hash-table-for-mapping-reply-serials-between-con.patch

```diff
@@ -0,0 +1,316 @@
+From e9b85276b6b382997197798ebc0071aa51e2d200 Mon Sep 17 00:00:00 2001
+From: Simo Piiroinen <simo.piiroinen@jolla.com>
+Date: Mon, 29 Nov 2021 15:33:51 +0200
+Subject: [PATCH] Use hash table for mapping reply serials between connections
+
+Using offsets for mapping serial numbers between bus and client
+connections can break down if proxy receives messages not meant for
+client consumption from bus while waiting for replies to method call
+messages sent by client. Which then leads to client side hitting
+timeout errors because replies get ignored due to incorrect reply
+serial numbers.
+
+Assign bus connection serials immediately after receiving data from
+client connection - so that bulk of proxy logic needs to deal only with
+bus connection serial numbers and offset calculations can be
+eliminated.
+
+Also - when dealing with method call messages - store bus to client
+connection mapping into hash table. Then utilize data in the hash table
+to modify reply serials when forwarding messages to client connection.
+
+Signed-off-by: Simo Piiroinen <simo.piiroinen@jolla.com>
+---
+ flatpak-proxy.c | 115 ++++++++++++++++++++++++++++++++----------------
+ 1 file changed, 76 insertions(+), 39 deletions(-)
+
+diff --git a/flatpak-proxy.c b/flatpak-proxy.c
+index 4282900..4156530 100644
+--- a/flatpak-proxy.c
++++ b/flatpak-proxy.c
+@@ -228,6 +228,7 @@ typedef struct
+   const char *sender;
+   const char *signature;
+   gboolean    has_reply_serial;
++  guint32     reply_serial_pos;
+   guint32     reply_serial;
+   guint32     unix_fds;
+ } Header;
+@@ -289,9 +290,10 @@ struct FlatpakProxyClient
+   ProxySide     bus_side;
+ 
+   /* Filtering data: */
+-  guint32     serial_offset;
+   guint32     hello_serial;
```

Comment:
Being able to remove those offset calculations is basically the point of this change...

But yeah, ddetecting reply to hello and triggering initial setup relies on "expecting reply" map rather than the serial. And that hello_serial field is not really used. So, it could be removed. Or one might use it for only-the-1st-hello type check (ignore rogue hello calls or stte).

Reply:
> Being able to remove those offset calculations is basically the point of this change...

Sure. I just thought that why it had to be detected by serial before but suddenly not anymore. I just didn't have the complete understanding of what this `hello_serial` actually used to mean. Was it like the very first serial, so no serial should be lower than that? Then it'd make a lot of sense that it's not needed now.

Reply:
After taking another look and a pinch of salt: git history does not cover when hello_serial appeared. But I have a feeling that it might have been about preserving hello reply_serial (always = 1?) despite there being proxy initiated ipc that would bump the offset up, so that only method calls *after* hello get their reply serials rewritten. Which sounds like one-off special case remedy that a more generic fix then renders unnecessary.

Reply:
After some pondering: I'll just drop the now unused hello_serial field instead of trying to find new use for it.